components/renku_data_services/solr/solr_migrate.py

```diff
@@ -0,0 +1,209 @@
+    migrationsRun: int
+    migrationsSkipped: int
+    requiresReindex: bool
+
+    @classmethod
+    def empty(cls) -> "MigrateResult":
+        """Create an empty MigrateResult."""
+        return MigrateResult(None, None, 0, 0, False)
+
+
+class VersionDoc(BaseModel):
+    """A document tracking the schema migration."""
+
+    id: str = "VERSION_ID_EB779C6B-1D96-47CB-B304-BECF15E4A607"
+    current_schema_version_l: int
+    migration_running_b: bool
+    version: int = pydantic.Field(
+        serialization_alias="_version_", validation_alias=AliasChoices("version", "_version_")
+    )
+
+    def to_dict(self) -> dict[str, Any]:
+        """Return the dict representation of this document."""
+        return self.model_dump(by_alias=True)
+
+
+class MigrationState(BaseModel):
+    """A private class tracking intermediate schema changes per migration."""
+
+    solr_schema: CoreSchema
+    doc: VersionDoc
+    skippedMigrations: int
+
+
+class SchemaMigrator:
+    """Allows to inspect the current schema version and run schema migrations against a solr core."""
+
+    def __init__(self, cfg: SolrClientConfig) -> None:
+        self.__config = cfg
+        self.__docId: str = "VERSION_ID_EB779C6B-1D96-47CB-B304-BECF15E4A607"
+
+    async def current_version(self) -> int | None:
+        """Return the current schema version."""
+        async with DefaultSolrClient(self.__config) as client:
+            doc = await self.__current_version0(client)
+            if doc is None:
+                return None
+            else:
+                return doc.current_schema_version_l
+
+    async def __current_version0(self, client: DefaultSolrClient) -> VersionDoc | None:
+        """Return the current schema version document."""
+        resp = await client.get_raw(self.__docId)
+        docs = resp.raise_for_status().json()["response"]["docs"]
+        if docs == []:
+            return None
+        else:
+            return VersionDoc.model_validate(docs[0])
+
+    async def migrate(self, migrations: list[SchemaMigration]) -> MigrateResult:
+        """Run all given migrations, skipping those that have been done before."""
+        async with DefaultSolrClient(self.__config) as client:
+            initialDoc = await self.__current_version0(client)
+            if initialDoc is None:
+                initialDoc = VersionDoc(
+                    current_schema_version_l=-1, migration_running_b=False, version=DocVersion.not_exists.value
+                )
+            return await self.__doMigrate(client, migrations, initialDoc)
+
+    async def __doMigrate(
+        self, client: DefaultSolrClient, migrations: list[SchemaMigration], initialDoc: VersionDoc
+    ) -> MigrateResult:
+        logging.info(
+            "".join(
+                [
+                    f"Core {self.__config.core}: Found current schema version: ",
+                    f"{initialDoc.current_schema_version_l} using {self.__docId}",
+                ]
+            )
+        )
```

Comment:
```suggestion
        logging.info(
            f"Core {self.__config.core}: Found current schema version: "
            f"{initialDoc.current_schema_version_l} using {self.__docId}"
        )

```

Reply:
This doesn't work for me - it yells a `TypeError: not all arguments converted during string formatting`

Reply:
You need to have no commas at the end. I think you get the error because you have left the commas.

Reply:
I.e. This will fail:

```python
logging.info(
            f"Core {self.__config.core}: Found current schema version: ", # <<<<< commas at end
            f"{initialDoc.current_schema_version_l} using {self.__docId}", # <<<<< commas at end
        )
```

But without the commas it will be fine

```python
logging.info(
            f"Core {self.__config.core}: Found current schema version: "
            f"{initialDoc.current_schema_version_l} using {self.__docId}"
        )
```